File: packages/csharp-sdk/src/Wusul/Resources/AccessPasses.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using Wusul.Http;
using Wusul.Models;

namespace Wusul.Resources
{
    /// <summary>
    /// AccessPasses resource for managing access passes
    /// </summary>
    public class AccessPasses
    {
        private readonly WusulHttpClient _http;

        /// <summary>
        /// Creates a new AccessPasses resource instance
        /// </summary>
        /// <param name="httpClient">HTTP client instance</param>
        public AccessPasses(WusulHttpClient httpClient)
        {
            _http = httpClient;
        }

        /// <summary>
        /// Issue a new access pass
        /// </summary>
        /// <param name="request">Parameters for issuing the access pass</param>
        /// <returns>The created access pass with installation URL</returns>
        public async Task<AccessPass> IssueAsync(IssueAccessPassRequest request)
        {
            return await _http.PostAsync<AccessPass>("/v1/access-passes", request);
        }

        /// <summary>
        /// List access passes with optional filtering
        /// </summary>
        /// <param name="request">Optional parameters for filtering</param>
        /// <returns>Array of access passes</returns>
        public async Task<List<AccessPass>> ListAsync(ListAccessPassesRequest? request = null)
        {
            var sigPayload = new Dictionary<string, object>();

            if (request != null)
            {
                if (!string.IsNullOrEmpty(request.TemplateId))
                {
                    sigPayload["template_id"] = request.TemplateId;
                }
                if (!string.IsNullOrEmpty(request.State))
                {
                    sigPayload["state"] = request.State;
                }
            }

            return await _http.GetAsync<List<AccessPass>>("/v1/access-passes", sigPayload.Count > 0 ? sigPayload : null);
        }

        /// <summary>
        /// Update an access pass
        /// </summary>
        /// <param name="accessPassId">The ID of the access pass to update</param>
        /// <param name="request">Parameters for updating the access pass</param>
        /// <returns>The updated access pass</returns>
        public async Task<AccessPass> UpdateAsync(string accessPassId, UpdateAccessPassRequest request)
        {
            return await _http.PatchAsync<AccessPass>($"/v1/access-passes/{accessPassId}", request);
        }

        /// <summary>
        /// Suspend an access pass
        /// </summary>
        /// <param name="accessPassId">The ID of the access pass to suspend</param>
        /// <returns>Success response</returns>
        public async Task<SuccessResponse> SuspendAsync(string accessPassId)
        {
            return await _http.PostAsync<SuccessResponse>($"/v1/access-passes/{accessPassId}/suspend");
        }

        /// <summary>
        /// Resume a suspended access pass
        /// </summary>
        /// <param name="accessPassId">The ID of the access pass to resume</param>
        /// <returns>Success response</returns>
        public async Task<SuccessResponse> ResumeAsync(string accessPassId)
        {
            return await _http.PostAsync<SuccessResponse>($"/v1/access-passes/{accessPassId}/resume");
        }

        /// <summary>
        /// Unlink an access pass from the user's device
        /// </summary>
        /// <param name="accessPassId">The ID of the access pass to unlink</param>
        /// <returns>Success response</returns>
        public async Task<SuccessResponse> UnlinkAsync(string accessPassId)
        {
            return await _http.PostAsync<SuccessResponse>($"/v1/access-passes/{accessPassId}/unlink");
        }

        /// <summary>
        /// Delete an access pass permanently
        /// </summary>
        /// <param name="accessPassId">The ID of the access pass to delete</param>
        /// <returns>Success response</returns>
        public async Task<SuccessResponse> DeleteAsync(string accessPassId)
        {
            return await _http.PostAsync<SuccessResponse>($"/v1/access-passes/{accessPassId}/delete");
        }
    }
}

File: packages/java-sdk/src/main/java/io/wusul/sdk/resources/AccessPasses.java
package io.wusul.sdk.resources;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.wusul.sdk.http.HttpClient;
import io.wusul.sdk.models.AccessPass;
import io.wusul.sdk.models.AccessPassState;
import io.wusul.sdk.models.IssueAccessPassParams;
import io.wusul.sdk.models.SuccessResponse;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * AccessPasses resource for managing access passes.
 */
public class AccessPasses {
    private final HttpClient http;
    private static final ObjectMapper objectMapper = new ObjectMapper();

    public AccessPasses(HttpClient httpClient) {
        this.http = httpClient;
    }

    /**
     * Issue a new access pass.
     *
     * @param params Parameters for issuing the access pass
     * @return The created access pass with installation URL
     * @throws IOException if the request fails
     */
    public AccessPass issue(IssueAccessPassParams params) throws IOException {
        return http.post("/v1/access-passes", params, AccessPass.class);
    }

    /**
     * List access passes with optional filtering.
     *
     * @return Array of access passes
     * @throws IOException if the request fails
     */
    public List<AccessPass> list() throws IOException {
        return list(null, null);
    }

    /**
     * List access passes with optional filtering.
     *
     * @param templateId Filter by template ID
     * @param state Filter by state
     * @return Array of access passes
     * @throws IOException if the request fails
     */
    public List<AccessPass> list(String templateId, AccessPassState state) throws IOException {
        Map<String, Object> sigPayload = new HashMap<>();

        if (templateId != null) {
            sigPayload.put("template_id", templateId);
        }
        if (state != null) {
            sigPayload.put("state", state.getValue());
        }

        // Using a workaround for generic List types
        Object response = http.get("/v1/access-passes", sigPayload.isEmpty() ? null : sigPayload, Object.class);
        return objectMapper.convertValue(response, new TypeReference<List<AccessPass>>() {});
    }

    /**
     * Update an access pass.
     *
     * @param accessPassId The ID of the access pass to update
     * @param updateData The update data
     * @return The updated access pass
     * @throws IOException if the request fails
     */
    public AccessPass update(String accessPassId, Map<String, Object> updateData) throws IOException {
        return http.patch("/v1/access-passes/" + accessPassId, updateData, AccessPass.class);
    }

    /**
     * Suspend an access pass.
     *
     * @param accessPassId The ID of the access pass to suspend
     * @return Success response
     * @throws IOException if the request fails
     */
    public SuccessResponse suspend(String accessPassId) throws IOException {
        return http.post("/v1/access-passes/" + accessPassId + "/suspend", null, SuccessResponse.class);
    }

    /**
     * Resume a suspended access pass.
     *
     * @param accessPassId The ID of the access pass to resume
     * @return Success response
     * @throws IOException if the request fails
     */
    public SuccessResponse resume(String accessPassId) throws IOException {
        return http.post("/v1/access-passes/" + accessPassId + "/resume", null, SuccessResponse.class);
    }

    /**
     * Unlink an access pass from the user's device.
     *
     * @param accessPassId The ID of the access pass to unlink
     * @return Success response
     * @throws IOException if the request fails
     */
    public SuccessResponse unlink(String accessPassId) throws IOException {
        return http.post("/v1/access-passes/" + accessPassId + "/unlink", null, SuccessResponse.class);
    }

    /**
     * Delete an access pass permanently.
     *
     * @param accessPassId The ID of the access pass to delete
     * @return Success response
     * @throws IOException if the request fails
     */
    public SuccessResponse delete(String accessPassId) throws IOException {
        return http.post("/v1/access-passes/" + accessPassId + "/delete", null, SuccessResponse.class);
    }
}
